File: clients/libjabi/peripherals/uart.cpp
#include <cstring>
#include <libjabi/byteorder.h>
#include <libjabi/interfaces/interface.h>

namespace jabi {

#include <jabi/peripherals.h>
#include <jabi/peripherals/uart.h>

void Device::uart_set_config(int baud, int data_bits,
        UARTParity parity, UARTStop stop, int idx) {
    iface_dynamic_req_t req = {
        .msg = {
            .periph_id   = PERIPH_UART_ID,
            .periph_idx  = static_cast<uint16_t>(idx),
            .periph_fn   = UART_SET_CONFIG_ID,
            .payload_len = sizeof(uart_set_config_req_t),
            .payload     = {0},
        },
        .payload = std::vector<uint8_t>(sizeof(uart_set_config_req_t), 0),
    };

    auto args = reinterpret_cast<uart_set_config_req_t*>(req.payload.data());
    args->baud      = htole<uint32_t>(baud);
    args->data_bits = static_cast<uint8_t>(data_bits);
    args->parity    = static_cast<uint8_t>(parity);
    args->stop_bits = static_cast<uint8_t>(stop);

    iface_dynamic_resp_t resp = interface->send_request(req);
    if (resp.payload.size() != 0) {
        throw std::runtime_error("unexpected payload length");
    }
}

void Device::uart_write(std::vector<uint8_t> data, int idx) {
    iface_dynamic_req_t req = {
        .msg = {
            .periph_id   = PERIPH_UART_ID,
            .periph_idx  = static_cast<uint16_t>(idx),
            .periph_fn   = UART_WRITE_ID,
            .payload_len = static_cast<uint16_t>(data.size()),
            .payload     = {0},
        },
        .payload = data,
    };

    iface_dynamic_resp_t resp = interface->send_request(req);
    if (resp.payload.size() != 0) {
        throw std::runtime_error("unexpected payload length");
    }
}

std::vector<uint8_t> Device::uart_read(size_t len, int idx) {
    iface_dynamic_req_t req = {
        .msg = {
            .periph_id   = PERIPH_UART_ID,
            .periph_idx  = static_cast<uint16_t>(idx),
            .periph_fn   = UART_READ_ID,
            .payload_len = sizeof(uart_read_req_t),
            .payload     = {0},
        },
        .payload = std::vector<uint8_t>(sizeof(uart_read_req_t)),
    };

    auto args = reinterpret_cast<uart_read_req_t*>(req.payload.data());
    args->data_len = htole<uint16_t>(static_cast<uint16_t>(len));

    iface_dynamic_resp_t resp = interface->send_request(req);
    if (resp.payload.size() > len) {
        throw std::runtime_error("unexpected payload length");
    }
    return resp.payload;
}

};

File: clients/rust/src/peripherals/uart.rs
use crate::interfaces::InterfaceRequest;
use crate::Error;
use deku::{DekuContainerWrite, DekuUpdate, DekuWrite};

pub enum UARTParity {
    None = 0,
    Odd = 1,
    Even = 2,
    Mark = 3,
    Space = 4,
}

pub enum UARTStop {
    B0_5 = 0,
    B1 = 1,
    B1_5 = 2,
    B2 = 3,
}

enum Func {
    SetConfig = 0,
    Write = 1,
    Read = 2,
}

fn gen_req(f: Func, idx: usize, payload: Vec<u8>) -> Result<InterfaceRequest, Error> {
    let mut r = InterfaceRequest {
        periph_id: crate::InstID::UART as u16,
        periph_idx: idx as u16,
        periph_fn: f as u16,
        payload_len: 0,
        payload: payload,
    };
    r.update().map_or(Err(Error::PacketFormat), |_| Ok(r))
}

impl crate::Device {
    pub fn uart_set_config(
        &self,
        idx: usize,
        baud: u32,
        data_bits: u8,
        parity: UARTParity,
        stop: UARTStop,
    ) -> Result<(), Error> {
        #[derive(DekuWrite)]
        #[deku(endian = "little")]
        struct SetConfigRequest {
            baud: u32,
            data_bits: u8,
            parity: u8,
            stop_bits: u8,
        }
        let req = SetConfigRequest {
            baud,
            data_bits,
            parity: parity as u8,
            stop_bits: stop as u8,
        };
        let resp = self.send(&gen_req(Func::SetConfig, idx, req.to_bytes().unwrap())?)?;
        if resp.len() == 0 {
            Ok(())
        } else {
            Err(Error::PacketFormat)
        }
    }

    pub fn uart_write(&self, idx: usize, data: &[u8]) -> Result<(), Error> {
        #[derive(DekuWrite)]
        #[deku(endian = "little")]
        struct WriteRequest<'a> {
            data: &'a [u8],
        }
        let req = WriteRequest { data };
        let resp = self.send(&gen_req(Func::Write, idx, req.to_bytes().unwrap())?)?;
        if resp.len() == 0 {
            Ok(())
        } else {
            Err(Error::PacketFormat)
        }
    }

    pub fn uart_read(&self, idx: usize, len: usize) -> Result<Vec<u8>, Error> {
        #[derive(DekuWrite)]
        #[deku(endian = "little")]
        struct ReadRequest {
            data_len: u16,
        }
        let req = ReadRequest {
            data_len: len as u16,
        };
        let resp = self.send(&gen_req(Func::Read, idx, req.to_bytes().unwrap())?)?;
        Ok(resp)
    }
}
